llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -295,15 +296,95 @@ saveResultModules(std::vector<std::unique_ptr<Module>> &ResModules) {
   return Res;
 }
 
+// Each fallback device library corresponds to one bit in "require mask" which
+// is an unsigned int32. getDeviceLibBit checks which fallback device library
+// is required for FuncName and returns the corresponding bit. The corresponding
+// mask for each fallback device library is:
+// fallback-cassert:      0x1
+// fallback-cmath:        0x2
+// fallback-cmath-fp64:   0x4
+// fallback-complex:      0x8
+// fallback-complex-fp64: 0x10
+static uint32_t getDeviceLibBits(const std::string &FuncName) {
+
+  static constexpr uint32_t DeviceLibAssert = 0x1;
+  static constexpr uint32_t DeviceLibCmath = 0x2;
+  static constexpr uint32_t DeviceLibCmath64 = 0x4;
+  static constexpr uint32_t DeviceLibComplex = 0x8;
+  static constexpr uint32_t DeviceLibComplex64 = 0x10;
+  if (FuncName == "__devicelib_assert_fail") {
+    return DeviceLibAssert;
+  }
+  size_t Len =
+      sizeof(CmathDeviceLibFunctions) / sizeof(CmathDeviceLibFunctions[0]);
+  if (std::binary_search(CmathDeviceLibFunctions, CmathDeviceLibFunctions + Len,
+                         FuncName)) {
+    return DeviceLibCmath;
+  }
+  Len =
+      sizeof(Cmath64DeviceLibFunctions) / sizeof(Cmath64DeviceLibFunctions[0]);
+  if (std::binary_search(Cmath64DeviceLibFunctions,
+                         Cmath64DeviceLibFunctions + Len, FuncName)) {
+    return DeviceLibCmath64;
+  }
+  Len =
+      sizeof(ComplexDeviceLibFunctions) / sizeof(ComplexDeviceLibFunctions[0]);
+  if (std::binary_search(ComplexDeviceLibFunctions,
+                         ComplexDeviceLibFunctions + Len, FuncName)) {
+    return DeviceLibComplex;
+  }
+  Len = sizeof(Complex64DeviceLibFunctions) /
+        sizeof(Complex64DeviceLibFunctions[0]);
+  if (std::binary_search(Complex64DeviceLibFunctions,
+                         Complex64DeviceLibFunctions + Len, FuncName)) {
+    return DeviceLibComplex64;
+  }
+  return 0;
+}
+
+// For each device image module, we go through all functions which meets
+// 1. The function name has prefix "__devicelib_"
+// 2. The function has SPIR_FUNC calling convention
+// 3. The function is declaration which means it doesn't have function body
+static uint32_t getModuleReqMask(const std::unique_ptr<Module> &MPtr) {
```

Comment:
Why are you using unique_ptr here and in getDeviceLibReqMasks?

Reply:
Hi, @andykaylor 
The main function in sycl-post-link.cpp has created a "std::vector<std::unique_ptr<Module>> ResultModules" and getDeviceLibReqMasks is invoked in main function, so I just use "ResultModules" as input parameters to avoid any conversion.
For getModuleReqMask, unique_ptr<Module> is not necessary, I have updated the patch to use "const Module &".